File: leetcode/3sum.cpp
// https://leetcode.com/problems/3sum/

#include <cmath>
#include <string>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

int main()
{
  vector<int> nums{-1, 0, 1, 2, -1, -4};
  sort(nums.begin(), nums.end());
  int fixed;
  int i, j;
  int sum;
  vector<vector<int>> ans;
  if (nums.size() >= 3)
  {
    for (fixed = 0; fixed < nums.size() - 2; fixed++)
    {
      
      i = fixed + 1;
      j = nums.size() - 1;
      while (i < j)
      {
        sum = nums[i] + nums[j];
        if (sum == -(nums[fixed]))
        {
          vector<int> a{nums[fixed], nums[i], nums[j]};
          ans.push_back(a);
          while (i < j && nums[i] == a[1])
            i++;
          while (i < j && nums[j] == a[2])
            j--;
          // break;
        }
        else if (sum < (-nums[fixed]))
        {
          i++;
        }
        else
        {
          j--;
        }
      }

      while (fixed + 1 < nums.size() && nums[fixed + 1] == nums[fixed])
      {
        fixed++;
      }

    }
  }
  else
  {
    exit(0);
  }

  for (vector<int> vec : ans)
  {
    for (int x : vec)
    {
      cout << x << " ";
    }
    cout << endl;
  }
}

File: grind75_python/3sum.py
# https://leetcode.com/problems/3sum/

from typing import List


class Solution:
  def threeSum(self, nums: List[int]) -> List[List[int]]:
    ans = []
    nums.sort()
    for index in range(len(nums) - 2):
      if index == 0 or nums[index - 1] != nums[index]:
        l = index + 1
        r = len(nums) - 1
        while l < r:
          if nums[l] + nums[r] + nums[index] > 0:
            r -= 1
          elif nums[l] + nums[r] + nums[index] < 0:
            l += 1
          else:
            ans.append([nums[index], nums[l], nums[r]])
            l += 1
            r -= 1
            while l+1 < len(nums) and nums[l] == nums[l-1]:
              l += 1
            while r-1 > index and nums[r] == nums[r+1]:
              r -= 1
    return ans


s = Solution()
print(s.threeSum([-2, 0, 1, 1, 2]))
